File: java/src/tarehart/rlbot/steps/wall/MountWallStep.java
package tarehart.rlbot.steps.wall;

import tarehart.rlbot.AgentInput;
import tarehart.rlbot.AgentOutput;
import tarehart.rlbot.input.CarData;
import tarehart.rlbot.math.BallSlice;
import tarehart.rlbot.math.vector.Vector3;
import tarehart.rlbot.physics.ArenaModel;
import tarehart.rlbot.physics.BallPath;
import tarehart.rlbot.planning.SteerUtil;
import tarehart.rlbot.steps.Step;

import java.awt.*;
import java.util.Optional;

public class MountWallStep implements Step {
    public Optional<AgentOutput> getOutput(AgentInput input) {

        CarData car = input.getMyCarData();

        if (ArenaModel.isCarOnWall(car)) {
            // Successfully made it onto the wall
            return Optional.empty();
        }

        BallPath ballPath = ArenaModel.predictBallPath(input);
        if (!WallTouchStep.hasWallTouchOpportunity(input, ballPath)) {
            // Failed to mount the wall in time.
            return Optional.empty();
        }

        BallSlice ballMotion = ballPath.getMotionAt(input.time.plusSeconds(3)).orElse(ballPath.getEndpoint());
        Vector3 ballPositionExaggerated = ballMotion.getSpace().scaled(1.04); // This assumes the ball is close to the wall

        return Optional.of(SteerUtil.steerTowardGroundPosition(car, ballPositionExaggerated));
    }

    @Override
    public boolean canInterrupt() {
        return true;
    }

    @Override
    public String getSituation() {
        return "Mounting the wall";
    }

    @Override
    public void drawDebugInfo(Graphics2D graphics) {
        // Draw nothing.
    }
}

File: java/src/tarehart/rlbot/steps/wall/MountWallStep.kt
package tarehart.rlbot.steps.wall

import tarehart.rlbot.AgentInput
import tarehart.rlbot.AgentOutput
import tarehart.rlbot.physics.ArenaModel
import tarehart.rlbot.planning.SteerUtil
import tarehart.rlbot.steps.Step

import java.awt.*
import java.util.Optional

class MountWallStep : Step {

    override val situation = "Mounting the wall"

    override fun getOutput(input: AgentInput): AgentOutput? {

        val car = input.myCarData

        if (ArenaModel.isCarOnWall(car)) {
            // Successfully made it onto the wall
            return null
        }

        val ballPath = ArenaModel.predictBallPath(input)
        if (!WallTouchStep.hasWallTouchOpportunity(input, ballPath)) {
            // Failed to mount the wall in time.
            return null
        }

        val (space) = ballPath.getMotionAt(input.time.plusSeconds(3.0)).orElse(ballPath.endpoint)
        val ballPositionExaggerated = space.scaled(1.04) // This assumes the ball is close to the wall

        return SteerUtil.steerTowardGroundPosition(car, ballPositionExaggerated)
    }

    override fun canInterrupt(): Boolean {
        return true
    }

    override fun drawDebugInfo(graphics: Graphics2D) {
        // Draw nothing.
    }
}
